pkg/controller/nexus/update/tags.go

```diff
@@ -0,0 +1,167 @@
+//     it under the terms of the GNU General Public License as published by
+//     the Free Software Foundation, either version 3 of the License, or
+//     (at your option) any later version.
+//
+//     Nexus Operator is distributed in the hope that it will be useful,
+//     but WITHOUT ANY WARRANTY; without even the implied warranty of
+//     MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+//     GNU General Public License for more details.
+//
+//     You should have received a copy of the GNU General Public License
+//     along with Nexus Operator.  If not, see <https://www.gnu.org/licenses/>.
+
+package update
+
+import (
+	"fmt"
+	"github.com/heroku/docker-registry-client/registry"
+	"github.com/m88i/nexus-operator/pkg/logger"
+	"strconv"
+	"strings"
+	"time"
+)
+
+const (
+	communityNexusRegistry     = "https://registry.hub.docker.com"
+	communityNexusRepo         = "sonatype/nexus3"
+	tagParseFailureFormat      = "unable to parse tag \"%s\": %v"
+	unableToCheckUpdatesFormat = "Unable to check for updates: %v"
+)
+
+const ttl = time.Hour * 6
+
+var (
+	lastQuery    time.Time
+	latestMicros = make(map[int]string)
+	log          = logger.GetLogger("update")
+)
+
+// HigherVersion checks if thisTag is of a higher version than otherTag
+func HigherVersion(thisTag, otherTag string) (bool, error) {
+	thisMinor, err := getMinor(thisTag)
+	if err != nil {
+		return false, fmt.Errorf(tagParseFailureFormat, thisTag, err)
+	}
+	otherMinor, err := getMinor(otherTag)
+	if err != nil {
+		return false, fmt.Errorf(tagParseFailureFormat, otherTag, err)
+	}
+	if thisMinor != otherMinor {
+		return thisMinor > otherMinor, nil
+	}
+
+	thisMicro, err := getMicro(thisTag)
+	if err != nil {
+		return false, fmt.Errorf(tagParseFailureFormat, thisTag, err)
+	}
+	otherMicro, err := getMicro(otherTag)
+	if err != nil {
+		return false, fmt.Errorf(tagParseFailureFormat, otherTag, err)
+	}
+	return thisMicro > otherMicro, nil
+}
+
+// GetLatestMicro returns the most recent image tag within a minor (the "y" in "x.y.z").
+// If the minor was not found or if we never managed to fetch any tags, the second return value is false.
+func GetLatestMicro(minor int) (tag string, ok bool) {
+	if time.Since(lastQuery) > ttl {
+		fetchUpdates()
+	}
+	tag, ok = latestMicros[minor]
+	return
+}
+
+// GetLatestMinor returns the most recent minor (the "y" in "x.y.z").
+// If there were issues fetching the tags it returns an error.
+func GetLatestMinor() (int, error) {
+	if time.Since(lastQuery) > ttl {
+		fetchUpdates()
+	}
+	if len(latestMicros) == 0 {
+		return 0, fmt.Errorf("unable to fetch tags")
+	}
+
+	greatestMinor := 0
+	for minor := range latestMicros {
+		if minor > greatestMinor {
+			greatestMinor = minor
+		}
+	}
+	return greatestMinor, nil
+}
+
+func fetchUpdates() {
+	tags, err := getTags()
+	if err != nil {
+		log.Errorf(unableToCheckUpdatesFormat, err)
+		return
+	}
+	lastQuery = time.Now()
```

Comment:
We might need an interval if we failed to fetch. Since we can fetch all day long uninterruptedly. I'd say 10 minutes and stop after 10 tries. Then try 6 hours later (using the `lastQuery`  ttl).

Reply:
Good point, I had not considered that. However, this would only be tried every reconcile, which happen automatically every hour or if there are changes to the watched objects. Is it really that bad if we attempt to fetch it on every reconcile?

If we do implement this mechanism to control how often we try again in case of failure, are you thinking of keeping it within the same reconcile iteration? Or are you thinking of trying again only on the next reconcile AND if we didn't fail in the last 10 minutes?

I assume you mean the latter, keeps a reconcile from possibly lasting 100 minutes, but just making sure we're on the same page.

After the 10th failure we can just do `lastQuery = time.Now()` and reset the failure count.

Reply:
I think we can add a simple guard in this function in case the reconcile loop happens sequentially like 4/5 times within a short period. This would avoid keeping fetching the docker API. So I'd say to not fetch if it was an error in the last minute. This would do.

Reply:
Sounds like a reasonable middle ground, will do.